## Обзор данных

Загружаю библиотеку pandas:

In [1]:
import pandas as pd

Знакомлюсь с данными

In [2]:
data = pd.read_csv('/content/02.data.csv')
data.head(10)

children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   
5         0    -926.185831         27    высшее             0   
6         0   -2879.202052         43    высшее             0   
7         0    -152.779569         50   СРЕДНЕЕ             1   
8         2   -6929.865299         35    ВЫСШЕЕ             0   
9         0   -2188.756445         41   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   
5  гражданский брак                 1      M   компаньон     0  255763.565419   
6   женат / замужем                 0      F   компаньон     0  240525.971920   
7   женат / замужем                 0      M   сотрудник     0  135823.934197   
8  гражданский брак                 1      F   сотрудник     0   95856.832424   
9   женат / замужем                 0      M   сотрудник     0  144425.938277   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

Есть необходимость приведения столбца `education` к единому виду: информация указана заглавными буквами, строчными, с заглавной буквы. Если оставить данные в таком виде, то получим ошибочные данные на выходе.

Смотрю какие типы данных в датасете, количество строк. Уже тут можно увидеть пропущенные значения

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


Коррелляция

In [4]:
data.corr()

<ipython-input-4-c44ded798807>:1: FutureWarning: The default value of numeric_only in DataFrame.corr is deprecated. In a future version, it will default to False. Select only valid columns or specify the value of numeric_only to silence this warning.
  data.corr()


children  days_employed  dob_years  education_id  \
children          1.000000      -0.138354  -0.174358     -0.012695   
days_employed    -0.138354       1.000000   0.582643      0.080565   
dob_years        -0.174358       0.582643   1.000000      0.067467   
education_id     -0.012695       0.080565   0.067467      1.000000   
family_status_id -0.091284       0.005726  -0.069988      0.007876   
debt              0.018470      -0.047110  -0.069620      0.052906   
total_income      0.018220      -0.136648  -0.052911     -0.178885   

                  family_status_id      debt  total_income  
children                 -0.091284  0.018470      0.018220  
days_employed             0.005726 -0.047110     -0.136648  
dob_years                -0.069988 -0.069620     -0.052911  
education_id              0.007876  0.052906     -0.178885  
family_status_id          1.000000  0.020611     -0.009147  
debt                      0.020611  1.000000     -0.012475  
total_income             -0.009147 -0.012475      1.000000

Прослеживается только одна взаимосвязь - это возраст клиента с трудовым стажем. Остальные данные не зависят друг от друга. Некоторые поддаются обратной коррелляции. Например, чем больше у клиента детей, тем меньше трудовой стаж.

**В итоге:**
1. В датасете 11 столбцов и 21525 строк. Это объемная выборка для проведения исследования.
Согласно документации к данным:
* `children` — количество детей в семье
* `days_employed` — общий трудовой стаж в днях
* `dob_years` — возраст клиента в годах
* `education` — уровень образования клиента
* `education_id` — идентификатор уровня образования
* `family_status` — семейное положение
* `family_status_id` — идентификатор семейного положения
* `gender` — пол клиента
* `income_type` — тип занятости
* `debt` — имел ли задолженность по возврату кредитов
* `total_income` — ежемесячный доход
* `purpose` — цель получения кредита

2. Есть необходимость в правке данных столбца `education`.

3. Количество значений в столбцах различается. Значит, в данных есть пропущенные значения.

4. У данных слабая коррелляция.

## Предобработка данных

Привожу столбец `education` к единому виду

In [5]:
data['education'] = data['education'].str.lower()

Ищу убликаты

In [6]:
data.duplicated().sum()

71

Нашла 71. Удаляю их

In [7]:
data = data.drop_duplicates()

Смотрю пропуски

In [8]:
data.isna().sum()

children               0
days_employed       2103
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2103
purpose                0
dtype: int64

Часть клиентов не указала ежемесячный доход и трудовой стаж. Заполняю данные о доходах средним значением в зависимости от типа занятости.

In [9]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

Меняю тип данных в столбце `total_income` на целочисленный. Это ежемесячный доход, не важны сотые доли

In [10]:
data['total_income'] = data['total_income'].astype(int)

В трудовом стаже есть аномальные значения - заменяю их

In [11]:
data['days_employed'] = data['days_employed'].abs()

Заполняю пропуски  медианными значениями по каждого типа занятости `income_type`

In [12]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

In [13]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

Медианное значение трудового стажа  для каждого типа занятости `days_employed` в днях

In [14]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения.  Оставляю данные в таком виде  столбец не понадобится для исследования.

Изучаю уникальные значения столбца `children`

In [15]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

Удаляю аномальные значения

In [16]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

In [17]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

**В итоге:**
1. Корректировка типа данных.
2. Данные приведены к единому виду.
3. Заполнены пропущенные значения.
4. Удалены явные и неявные дубликаты.
5. Удалены аномалии.

### Категоризация данных

Напишу функцию для определения категорий по ежемесячному доходу:

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


In [18]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

Проверяю ее работу

In [19]:
categorize_income(112080)

'C'

Все верно! Клиент с доходом 112080  имеет категорию `С`.

Добавляю столбец `total_income_category` с категориями:

In [20]:
data['total_income_category'] = data['total_income'].apply(categorize_income)
display(data.head())

children  days_employed  dob_years education  education_id  \
0         1    8437.673028         42    высшее             0   
1         1    4024.803754         36   среднее             1   
2         0    5623.422610         33   среднее             1   
3         3    4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   

                      purpose total_income_category  
0               покупка жилья                     B  
1     приобретение автомобиля                     C  
2               покупка жилья                     C  
3  дополнительное образование                     B  
4             сыграть свадьбу                     C

Смотрю какие есть уникальные цели для взятия кредита из столбца `purpose`

In [21]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

Напишу функцию для определения целей кредита. Судя по уникальным значениям, можно выделить категории:

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

In [22]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

Проверяю ее работу

In [23]:
categorize_purpose('автом')

'операции с автомобилем'

Отлично! При вводе корня `автом` клиент попадает под категорию `операции с автомобилем`.

Добавляю `purpose_category`,  с категориями

In [24]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)
display(data.head())

children  days_employed  dob_years education  education_id  \
0         1    8437.673028         42    высшее             0   
1         1    4024.803754         36   среднее             1   
2         0    5623.422610         33   среднее             1   
3         3    4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   

                      purpose total_income_category          purpose_category  
0               покупка жилья                     B  операции с недвижимостью  
1     приобретение автомобиля                     C    операции с автомобилем  
2               покупка жилья                     C  операции с недвижимостью  
3  дополнительное образование                     B     получение образования  
4             сыграть свадьбу                     C        проведение свадьбы

#### Исследование данных


 <u>Есть ли зависимость между количеством детей и возвратом кредита в срок?</u>

In [25]:
data_pivot = data.pivot_table(index = ['children'], values = 'debt', aggfunc = ['count', 'mean', lambda x: round(len(x)/len(data)*100,2)])
display(data_pivot)

count      mean <lambda>
           debt      debt     debt
children                          
0         14091  0.075438    66.06
1          4808  0.092346    22.54
2          2052  0.094542     9.62
3           330  0.081818     1.55
4            41  0.097561     0.19
5             9  0.000000     0.04

Зависимость возврата кредита от количества детей есть.

В ходе исследования данных оказалось, что кредит чаще всего берут бездетные граждане и не выплачивают вовремя в 7,5%. Общая доля от количества должников составляет 66%.

Безответственная категория плательщиков - граждане имеющие 1-2 ребёнка. Вовремя не оплачивают кредит примерно 9,3% случаях. Общая доля от количества должников составляет 22%.

Граждан с 3-5 детьми предлагаю не рассматривать. Выборка меньше 2%, то есть она не показательна.

<u>Есть ли зависимость между семейным положением и возвратом кредита в срок?</u>

In [26]:
data_pivot = data.pivot_table(index = ['family_status'], values = 'debt', aggfunc = ['count', 'mean', lambda x: round(len(x)/len(data)*100,2)])
display(data_pivot)

count      mean <lambda>
                        debt      debt     debt
family_status                                  
Не женат / не замужем   2796  0.097639    13.11
в разводе               1189  0.070648     5.57
вдовец / вдова           951  0.066246     4.46
гражданский брак        4134  0.093130    19.38
женат / замужем        12261  0.075606    57.48

Зависимость возврата кредита от семейного положения есть.

В ходе исследования данных оказалось, что кредит чаще всего берут семейные граждане и не выплачивают его вовремя в 7,5%. Они составляют самую большую категорию кредитников, 57%. Если учитывать данные исследования о наличии детей, также можно предположить, что это бездетные семьи.

Безответственные категории плательщиков - свободные граждане/гражданский брак. Платят позже срока примерно 9,5%. Общая доля от количества должников составляет 32%.

Граждане в разводе/вдовцы не выполняют вовремя обязательства по кредиту всего примерно 7%. Но их общая доля от количества должников составляет всего 10%.

<u>Есть ли зависимость между уровнем дохода и возвратом кредита в срок?</u>

In [27]:
data_pivot = data.pivot_table(index = ['total_income_category'], values = 'debt', aggfunc = ['count', 'mean', lambda x: round(len(x)/len(data)*100,2)])
display(data_pivot)

count      mean <lambda>
                        debt      debt     debt
total_income_category                          
A                         25  0.080000     0.12
B                       5014  0.070602    23.51
C                      15921  0.084982    74.64
D                        349  0.060172     1.64
E                         22  0.090909     0.10

Зависимость возврата кредита от уровня дохода есть.
Граждане с доходом из категории `С` не возвращают кредит вовремя почти в 8,5%. Это самая большая группа и она составляет 74% от общего количества должников.
Граждане с доходом категории `B` не выплачивают вовремя 7%. Это вторая основная группа должников, она составляет 23%.

Доля должников категории В ниже должников категории `С` в три раза.

Категории уровня дохода:
* 0–30000 — `E`
* 30001–50000 — `D`
* 50001–200000 — `C`
* 200001–1000000 — `B`
* 1000001 и выше — `А`.


<u>Как разные цели кредита влияют на его возврат в срок?</u>

In [28]:
data_pivot = data.pivot_table(index = ['purpose_category'], values = 'debt', aggfunc = ['count', 'mean', lambda x: round(len(x)/len(data)*100,2)])

display(data_pivot)


count      mean <lambda>
                           debt      debt     debt
purpose_category                                  
операции с автомобилем     4279  0.093480    20.06
операции с недвижимостью  10751  0.072551    50.40
получение образования      3988  0.092528    18.70
проведение свадьбы         2313  0.079118    10.84

Проще всего даётся погасить кредит вовремя гражданам, которые покупают жилье: всего 7,2%. Это самая большая доля - 50% от общего количества должников.

При покупке автомобиля и повышение образования - 9,3%. Доля: 20% и 19% соответственно. Самая безответственная категория.

У тех, кто не накопил на свадьбу доля - 11%. Не отдают кредит вовремя 7,9%.

### Выводы

### Исследование данных показало:
* Самыми ответственными плательщиками по кредиту являются бездетные женатые граждане. Учитывая, что доля категории покупки жилья самая большая, можно предположить, что это молодые семьи, берущие ипотеку. Они наша целевая аудитория
* С осторожностью давать кредиты более безответственным плательщикам: это граждане с 1-2 детьми, граждане без официальной пары. Возможно, пересмотреть условия кредитования.
* Малоимущие и очень обеспеченные граждане редко обращаются за кредитами.